File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/Y17D05.kt
package org.gristle.pdxpuzzles.advent.y2017

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y17D05(input: String) : Day {
    private val jumps = input.lines().map(String::toInt)

    private inline fun solve(increment: (Int) -> Int): Int {
        val offsets = jumps.toIntArray()
        var i = 0
        var steps = 0
        while (i < offsets.size) {
            steps++
            i += offsets[i].also { offsets[i] += increment(offsets[i]) }
        }
        return steps
    }

    override fun part1(): Int = solve { 1 }

    override fun part2(): Int = solve { if (it >= 3) -1 else 1 }
}

fun main() = Day.runDay(Y17D05::class)

//    Class creation: 22ms
//    Part 1: 373160 (29ms)
//    Part 2: 26395586 (275ms)
//    Total time: 327ms

File: rust/advent/src/bin/2017/y17d05.rs
use advent::utilities::get_input::get_input;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input = Vec<i64>;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(17, 5).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.get_numbers().collect()
}

fn solve(jumps: &Input, increment: fn(i64) -> i64) -> Output {
    let mut offsets: Vec<i64> = jumps.iter().copied().collect();
    let mut i = 0;
    let mut steps = 0;
    while i < offsets.len() {
        steps += 1;
        let old_i = i;
        i = (i as i64 + offsets[i]) as usize;
        offsets[old_i] += increment(offsets[old_i])
    }
    steps
}

fn part1(jumps: &Input) -> Output {
    solve(jumps, |_| 1)
}

fn part2(jumps: &Input) -> Output {
    solve(jumps, |it| if it >= 3 { -1 } else { 1 })
}

#[test]
fn default() {
    let input = get_input(17, 5).unwrap();
    let input = parse_input(&input);
    assert_eq!(373160, part1(&input));
    assert_eq!(26395586, part2(&input));
}

// Input parsed (261μs)
// 1. 373160 (512μs)
// 2. 26395586 (35ms)
// Total: 36ms
